Refactorización inicial

devarops opened the pull request:
@efren-cabrera: ¿Tienes alguna sugerencia sobre cómo debo refactorizar lo que llevamos?

devarops committed:
Agrega versión inicial

devarops committed:
Agrega archivos de configuración

efren-cabrera commented on src/tablero.py:
```diff
@@ -0,0 +1,45 @@
+    tablero_ramas = registro_ramas[es_rama]
+    tablero = pd.DataFrame(columns=['repo', 'objetivo', 'develop', 'default'])
+    for (repo, objetivo), tabla in tablero_ramas.groupby(by=["repo","objetivo"]):
+        es_develop = ((tabla.revision == "develop") | (tabla.revision == "development"))
```
Esto lo repites en la línea 25

devarops replied:
Lo regactoricé con una funcioncita de un renglón: `is_develop`

efren-cabrera commented on src/tablero.py:
```diff
@@ -0,0 +1,45 @@
+    registro_ramas = pd.DataFrame(
+        columns=['repo', 'objetivo', 'revision', 'exitoso'])
+    for (repo, objetivo, revision), tabla in registro_testmake.groupby(["repo", "objetivo", "revision"]):
+        ultimo_renglon = tabla.iloc[-1]
```
Tal vez aquí sea necesario agregar el comentario de por qué buscas el último registro (el más nuevo)

devarops replied:
Agregué un comentario; espero que sea claro

efren-cabrera commented on src/tablero.py:
```diff
@@ -0,0 +1,45 @@
+
+def get_records(log_name="~/.testmake/testmake.log.csv"):
+    registro_testmake = pd.read_csv(log_name)
+    registro_testmake.timestamp = pd.to_datetime(registro_testmake.timestamp)
```
¿es necesario transformarlos en fecha?
si están en ISO-8601  estas tres líneas podrían ser una sola

`registro_testmake = pd.read_csv(log_name).sort_values(by="timestamp")`

devarops replied:
No sé si es necesario; no sé como _debuggiar_. Necesitaré de tu ayuda...

devarops replied:
Listo.

efren-cabrera commented on src/templates/index.html:
```diff
@@ -0,0 +1,27 @@
+
+            {% for registro in tablero %}
+            <tr>
+                    <td>{{registro[0]}}</td>
```
Las identaciones de este archivo están mal

devarops replied:
Creo que ya quedó. Le pedí al VS Code que lo formatee.

efren-cabrera reviewed (changes requested):
Aquí va una pequeña tanda de correcciones, aún no termino. Me gustaría que cambiaramos la estructura del repositorio. Una propuesta rápida que se me ocurre ahorita sería algo así

```
├── app.py
├── tablero
│      ├── __init__.py
│      ├── [sumodulo para leer archivo o descargar los datos de una api, tal vez io o data]
│              ├── __ini__.py
│              ├── load_record.py (aquí podrías poner la función get_records)
│      ├── Dashboard.py (Aquí creas una clase que represente al dashboard, esto te conviene para no mandar una lista al front end, mejor un json, si tenemos el objeto es fácil pasarlo a json, tal vez que el constructor acepte la salida de get_records y después se pueda sustituir por otra cosa, sería el equivalente al get_dashboard)
├── templates
│      ├── index.html
├── setup.py
├── .gitignore
```

Creo que es buena práctica siempre dejar una línea en blanco al final del archivo.

devarops commented:
Gracias por tu revisión, @efren-cabrera. Hacerlo por tandas me parece una buena idea. Iré incorporando esta primer tanda.

devarops committed:
Corrige las identaciones

devarops committed:
Cabia la estructura del repositorio

devarops committed:
Agrega línea vacía al final de cada archivo

devarops committed:
Describe el propósito de la aplicación

devarops committed:
Agrega comentario explicando por qué por qué quiero el último registro

También formatea el archivo

devarops committed:
Considera nueva estructura al importar función `get_dashboard()`

devarops committed:
Formatea tabla usando negritas para repos y monoespaciado para objetivos

devarops committed:
Usa última ocurrencia en lugar de primer ocurrencia

por que la última ocurrencia es la más reciente

devarops committed:
Deja de trasforma a fecha por que no es necesario para ordenar

ya que timestam está en están en ISO-8601

devarops committed:
Agrega línea para _debugging_

devarops committed:
Importa función `get_records` en submódulo `io`

y formatea archivo

devarops commented:
@efren-cabrera: Me faltó hacer la clase `Dashboard` pero puedes ir revisando el resto

efren-cabrera commented on tablero/Dashboard.py:
```diff
@@ -0,0 +1,41 @@
+    tablero = pd.DataFrame(columns=['repo', 'objetivo', 'develop', 'default'])
+    for (repo, objetivo), tabla in tablero_ramas.groupby(by=["repo", "objetivo"]):
+        es_develop = is_develop(tabla.revision)
+        if not np.any(es_develop):
```
La línea 25 a la 30  se [arece mucho a la línea 32 a 37 ¿se podría hacer una función para reducir el tamaño de esta función y evitar código repetido?

devarops replied:
Tienes razón. Así lo haré.

devarops replied:
Creé la función `is_successful()`

efren-cabrera reviewed (changes requested):
Aquí te va otra sugerencia

efren-cabrera commented on tablero/io/load_record.py:
```diff
@@ -0,0 +1,22 @@
+        # `tabla` contiene los registros de todos los intentos de generar el
+        # objetivo, de los cuales sólo nos interesa el más reciente que guardo
+        # en `ultimo_renglon`.
+        ultimo_renglon = tabla.iloc[-1]
```
¿Qué opinas si en vez de poner un comentario tan largo haces una función? algo como `get_last_try_status` y en la documentación de la función puede ir este comentario

devarops replied:
OK. Así lo hare.

devarops replied:
Creé la función `get_last_record()`

devarops committed:
Crea función `is_successful()` para evita código repetido

siguiendo [esta sugerencia](https://github.com/IslasGECI/tablero/pull/1#discussion_r304177468)

devarops committed:
Crea función `get_last_record()` para encapsular instrucción críptica

y ofrecer un comentario verboso siguiendo [esta sugerencia](https://github.com/IslasGECI/tablero/pull/1#discussion_r304178068)

devarops commented:
@efren-cabrera: Ya atendí la segunda tanda de comentarios.
Sólo me falta hacer la clase `Dashboard`.
¿Cómo te imaginas que se vería el comando `render_template("index.html", tablero=get_dashboard())` cuando haya creado la clase `Dashboard`?

efren-cabrera commented on app.py:
```diff
@@ -0,0 +1,11 @@
+app = Flask(__name__)
+
+@app.route('/')
+def hello_world():
```
¿Podrías cambiar el nombre de esta función?, tal vez podría ser index o dashboard

efren-cabrera replied:
@evaristor Este es el único cambio que queda pendiente para que te la apruebe

efren-cabrera commented:
> @efren-cabrera: Ya atendí la segunda tanda de comentarios.
> Sólo me falta hacer la clase `Dashboard`.
> ¿Cómo te imaginas que se vería el comando `render_template("index.html", tablero=get_dashboard())` cuando haya creado la clase `Dashboard`?

@evaristor Tal vez la clase la podemos dejar para después. Podríamos crear dos clases una de `Record` o `Repo` que guarde el nombre del repo, nombre del reporte o resultado, el status de la rama _default_ y el de la rama _develop_.

Yo me imagino que podrías hacer algo como

``` python
class Record:
    def __init__(self, repositorie_name, report_name, status_default, status_develop):
        self.repositorie_name = repositorie_name
        ...

class Dashboard:
    __init__(self, log_name="~/.testmake/testmake.log.csv"):
        self.log_name = log_name
        self._records = self.read_log_file()

    def read_log_file(self) -> List[Records]:
        # Aquí iría la mayoría del código de `get_records`
        ...

    @property
    def records():
        return self._records
```

Y ya en la aplicación podríamos utilizarlo

```python
@app.route('/')
def dashboard():
    dashboard = Dashboard()
    return render_template("index.html", tablero=dashboard)
```

Y en el _tamplate_ tendríamos que poner

```python
{% for registro in tablero.records %}
    <tr>
        <td><b>{{registro.repositorie_name}}</b></td>
        <td><code>{{registro.report_name}}</code></td>
        <td><img src="{{registro.status_default}}" /></td>
        <td><img src="{{registro.status_develop}}" /></td>
    </tr>
{% endfor %}
```

devarops committed:
Renombra dunción a index en lugar de nombre de la plantilla hello_wolrd
